Use a DVV like approach to stop sibling explosion

russelldb opened the pull request:
As has been witnessed by our customers, and demonstrated by Baquero et al, sibling explosion occurs when we keep values as concurrent when they're not really.

Our vnode vclocks are not fine grained enough to describe which of a set of siblings should be overwritten by a PUT, so we keep all siblings. It is trivially easy to create a scenario where interleaving writes create an explosion of sibling values, even without failure, or concurrency (see basho/riak_test#383.)

This PR introduces a more fine grained approach to causality tracking, without going all the way and using DVV or DVVSets.

The general mechanism is to store the event (`{actor, {count, ts}}`) that a value is created at as metadata with the value. When a PUT is received, any siblings whose events are dominated by the PUT vclock are removed. Only siblings that the writer has not seen are genuinely concurrent.

In order not to drop ALL the values during other merge types, when any two merging objects both have the same event, that sibling stays.

Nothing is for free, so there is some extra complexity around `riak_object:merge`.

This PR requires the basho/riak_core#463 and can be tested with basho/riak_test#480.

russelldb committed:
Stop sibling explosion

Add an event `dot` to metadata on PUT, use this to drop
siblings that are dominated by a put.

russelldb committed:
Remove test for removed code

russelldb committed:
Re-instate the removal of dominated and duplicate objects

Fix ec_eqc to use the DVV logic and expect fewer siblings

russelldb committed:
Re-add test (since I re-added the thing it tests)

russelldb committed:
Add tests for mixed dotted an undotted merge

Data written before dots must merge safely with dotted data
a value with no dot os always kept.

russelldb committed:
Add "killswitch" for DVV

Harden code a little

russelldb committed:
Add `dvv` option to cuttlefish

If `dvv` is `off` then use the legacy merge function which does
not use `dots`

russelldb committed:
Merge CRDTs inside riak_object merge

Why? Well it means no CRDT siblings on disk. Previously CRDTs
were merged only when a CRDT Operation was performed from the CRDT
API endpoints. This means read repair, AAE, and repl, could all cause
siblings on disk.

With the addition of DVV style causality, we need to make a choice
about CRDTs. We don't have DVVSets and the anonymous list. So CRDTs
either have to merge Dots into a mini-version-vector, or not have dots.

If CRDTs don't have dots, so their values always survive a merge,
then they are subject to repl induced sibling-explosion (possibly)

Merging in riak_object:merge solves that issue.
As an added benefit it also means that Yokozuna indexing of CRDTs
will just work.

I'd really appreciate a performance comparision between this and
the plain DVV branch, and the pre-DVV stuff for 2.0.

russelldb committed:
Merge pull request #758 from basho/bug/rdb/sibling-explosion-crdt

Merge CRDTs inside riak_object merge

russelldb commented on src/riak_object.erl:
```diff
@@ -387,12 +505,52 @@ set_vclock(Object=#r_object{}, VClock) -> Object#r_object{vclock=VClock}.
+-spec assign_dot(riak_object(), vclock:dot(), boolean()) -> riak_object().
+assign_dot(Object, Dot, true) ->
+    #r_object{contents=[C=#r_content{metadata=Meta0}]} = Object,
+    %% Dot must actually be a vclock, allbeit it one with a single entry
```
oops, this comment was left by mistake, and needs removing

Vagabond commented on src/riak_object.erl:
```diff
@@ -239,30 +263,124 @@ compare_content_dates(C1,C2) ->
+    MergeAcc0 = prune_object_siblings(OldObject, vclock(NewObject)),
+    MergeAcc = prune_object_siblings(NewObject, vclock(OldObject), MergeAcc0),
+    #merge_acc{crdt=CRDT, error=Error} = MergeAcc,
+    riak_kv_crdt:log_merge_errors(Bucket, Key, CRDT, Error),
```
Do we consider a classic riak_object a CRDT, or are we just piggybacking on handy functions from the CRDT module?

russelldb replied:
Previously, we merged CRDTs in the vnode on a CRDT Operation PUT only. This means CRDT siblings can be generated by repl, AAE, and Read Repair. CRDT values can't be dotted: There is no single PUT to overwrite previous values, instead a merge of values generates  a new value. In DVVSets they have the "anonymous list" for these generated values. In Riak it makes sense to have them undotted, as undotted siblings always survive. However, if CRDTs are undotted, then they can "explode" via Repl. To avoid this, we merge crdts in riak_object merge. It has other advantages too (mainly that yokozuna can index CRDTs). It is also a slight optimization on the CRDT path. The main thing is that this PR has DVV _and_ riak CRDT merging in riak_object merge.

russelldb replied:
Oops, the code you refer to just logs any CRDT merge issues encountered during the merge operation. It reuses the existing CRDT code, yes.

Vagabond commented on src/riak_object.erl:
```diff
@@ -168,12 +182,18 @@ equal_contents([C1|R1],[C2|R2]) ->
 ancestors(Objects) ->
-    ToRemove = [[O2 || O2 <- Objects,
-                       vclock:descends(O1#r_object.vclock,O2#r_object.vclock),
-                       (vclock:descends(O2#r_object.vclock,O1#r_object.vclock) == false)]
```
This is subtly different than the vclock:dominates() call, in that dominates() checks if the vclocks are equal. This is fine, right?

russelldb replied:
They're the same, the only way two vclocks could descend each other is if they were equal. I think `vclock:dominates/2` makes that clearer (although, IMO, equals check first might be an optimisation.)

seancribbs replied:
Suggested simplification (multiple generators are same as nested comprehensions):

``` erlang
[ O2 || O1 <- Objects, O2 <- Objects, strict_descendant(O1, O2) ]
```

seancribbs replied:
If you chose the simplification, you also don't have to flatten, since it generates a flat list already.

Vagabond commented on src/riak_object.erl:
```diff
@@ -182,22 +202,16 @@ ancestors(Objects) ->
 %%       X-Riak-Last-Modified header.
 -spec reconcile([riak_object()], boolean()) -> riak_object().
 reconcile(Objects, AllowMultiple) ->
```
I'm getting a little lost here. It would be nice to have comments for dedup and reconcile/2 and reconcile2/2. I sort of understand it, but would like to feel more confident.

russelldb committed:
Remove old dedupe logic as it is no longer needed

The new merge function by CMeik actually removes the need for
object deduplication (it deduplicates contents).

Well spotted Vagabond!

russelldb commented:
@Vagabond pushed a commit that removes dedupe logic and adds comments.

Vagabond commented on src/riak_object.erl:
```diff
@@ -182,34 +202,42 @@ ancestors(Objects) ->
+reconcile([Object1, Object2 | Rest]) ->
+    %% Start reconciling merging the first two objects in the list to
+    %% get a most merged object.
+    reconcile2(Rest, syntactic_merge(Object1, Object2)).
```
Maybe something like this would be clearer than reconcile2/2?

``` erlang
lists:foldl(fun(Obj, MergeDest) ->
    syntactic_merge(Obj, MergeDest)
end, hd(Objects), tl(Objects))
```

russelldb replied:
except the single object in a list case?

Vagabond replied:
that should work:

``` erlang
lists:foldl(fun(E, A) -> E + A end, hd([1]), tl([1])).
1
```

But I meant to replace reconcile2/2 with it, I typoed originally.

russelldb replied:
Okay, will try that. I consider it stylistic, but if it makes it easier to read, then that is good.

seancribbs replied:
Single-object list case will still work in @Vagabond's version.

``` erlang
lists:foldl(fun syntactic_merge/2, hd(Objects), tl(Objects))
```

This is the definition of lists:foldl:

``` erlang
foldl(F, Accu, [Hd|Tail]) ->
    foldl(F, F(Hd, Accu), Tail);
foldl(F, Accu, []) when is_function(F, 2) -> Accu.
```

russelldb replied:
Sorry, I forgot the order of the args (it is Fun, Acc, List, I thought it was Fun, List, Acc, ('cos the Fun takes Element, Acc (which is more like the order List, Acc))) I blame the lists lib :D

seancribbs replied:
Indeed. My rule of thumb is "list last".

evanmcc commented:
these are n=1 numbers:
develop:
get 1 avg = 3264, gets [2865,3477,3452]
get 128 avg = 46938, gets [48297,44438,48079]
put 1 avg = 1797, puts [1788,1827,1778]
put 128 avg = 32391, puts [32608,32407,32158]

dvv @ f00aac8
get 1 avg = 4108, gets [4137,4032,4157]
get 128 avg = 56012, gets [56413,55948,55677]
put 1 avg = 1709, puts [1670,1644,1814]
put 128 avg = 33558, puts [33621,33648,33405]

dvv @ dcca904:
get 1 avg = 4090, gets [4216,4116,3938]
get 128 avg = 56423, gets [56834,56278,56158]
put 1 avg = 1806, puts [1816,1842,1760]
put 128 avg = 33463, puts [33835,33357,33198]

Vagabond commented on src/riak_object.erl:
```diff
@@ -239,30 +267,127 @@ compare_content_dates(C1,C2) ->
+        {ok, Dot} ->
+            case {vclock:descends(Clock, [Dot]), orddict:is_key(Dot, Drop)} of
+                {true, true} ->
+                    %% Dot present on both sides, keep it
```
I know we talked about this logic, but more comments here about this logic would be nice.

Vagabond commented on src/riak_object.erl:
```diff
@@ -239,30 +267,127 @@ compare_content_dates(C1,C2) ->
+merge_acc_to_contents(MergeAcc) ->
+    #merge_acc{keep=Keep, crdt=CRDTs} = MergeAcc,
+    Keep2 = lists:map(fun convert_from_dict_list/1, ordsets:to_list(Keep)),
+    orddict:fold(fun(_Type, {Meta, CRDT}, Acc) ->
```
_Type is always r_content, right? Should we use Type on the next line instead of having that r_content literal, or pattern match on r_content so we can't accidentally strip type information if more types are added later?

russelldb replied:
No. Type is a CRDT type (riak_dt_pncounter, riak_dt_map, riak_dt_orswot etc). The orddict that is being mapped over is a mapping of Type-> {Meta, CRDT value}. We don't need the type in this fold, so we don't use it.

Vagabond commented on src/riak_object.erl:
```diff
@@ -239,30 +267,127 @@ compare_content_dates(C1,C2) ->
+    Keep2 = lists:map(fun convert_from_dict_list/1, ordsets:to_list(Keep)),
+    orddict:fold(fun(_Type, {Meta, CRDT}, Acc) ->
+                         [{r_content, riak_kv_crdt:meta(Meta, CRDT),
+                           riak_kv_crdt:to_binary(CRDT)} | Acc]
```
I'm assuming that for regular riak_objects, the 'mod' in the CRDT record is riak_object? All these CRDT references in riak_object are kind of freaking me out.

russelldb replied:
Regular riak objects are not gathered in the orddict. They're dealt with separately.

russelldb replied:
Can you explain your issue with CRDT code in riak_object? We have CRDTs now. There is no pure separation between riak_object and CRDTs. The merging of contents is what this function does. CRDT just means "non-opaque content item". Merge has extended what it does. I don't understand the issue. Is it some "law of demeter" thing, or a general squeamish-ness? We could merge the riak object contents, then pass the object to riak_kv_crdt and merge it again. The munging has occurred for optimisation reasons (like the comments says.) If we're iterating over the contents, don't do it again in riak_kv_crdt.  This commit comment, hopefully explains it well enough https://github.com/basho/riak_kv/commit/18fae0511b56ee9e65dcb74c21c401e44a7e32e8

Vagabond commented on src/riak_object.erl:
```diff
@@ -239,30 +267,127 @@ compare_content_dates(C1,C2) ->
+%% @private Get the dot from the passed metadata dict
+%% (if present and valid).
+-spec get_dot(dict()) -> {ok, vclock:dot()} | undefined.
+get_dot(Dict) ->
```
Are there known cases when either the dot is not valid, or is not present in the vclock? If so, can we document them? The only time I can think of is if a vclock is pruned...?

russelldb replied:
None that I can think of. Do you think we should a) log here or b) just let it crash at the point the dot is compared to the clock? c) crash here?

Vagabond commented on src/riak_object.erl:
```diff
@@ -239,30 +267,127 @@ compare_content_dates(C1,C2) ->
 %%      ensure we can do comparsions between r_contents.
 convert_to_dict_list({r_content, Dict, Value}) ->
+    {r_content, lists:usort(dict:to_list(Dict)), Value};
+convert_to_dict_list({Dict, Value}) ->
```
When do we get untyped objects like this?

russelldb replied:
Never, it is from a previous version. I'll remove it.

Vagabond commented on src/riak_object.erl:
```diff
@@ -959,14 +1127,6 @@ is_updated_test() ->
     OVu = riak_object:update_value(O, testupdate),
     ?assert(is_updated(OVu)).
 
-remove_duplicates_test() ->
```
Was this test removed for a reason?

russelldb replied:
The code it tests was removed.

Vagabond commented on src/riak_object.erl:
```diff
@@ -1036,4 +1196,52 @@ vclock_codec_test() ->
+    ?assertEqual(2, vclock:get_counter(z, Vclock)),
+    ?assertEqual(3, length(Vclock)).
+
+mixed_merge_test() ->
```
The mixed_merge tests could use some comments to clear up what we're actually testing here.

Vagabond commented on src/riak_object.erl:
```diff
@@ -387,12 +512,52 @@ set_vclock(Object=#r_object{}, VClock) -> Object#r_object{vclock=VClock}.
-    Object#r_object{vclock=vclock:increment(ClientId, Object#r_object.vclock)}.
+    NewClock = vclock:increment(ClientId, Object#r_object.vclock),
+    {ok, Dot} = vclock:get_entry(ClientId, NewClock),
+    assign_dot(Object#r_object{vclock=NewClock}, Dot, dvv_enabled()).
```
calling assign_dot/3 here instead of assign_dot/2 skips the vclock:valid_entry check, is this intentional?

russelldb replied:
Yes. The comment explains it. assign/2 is an exported function, so it can't trust the dot. assign/3 is called inside riak_object, and only with a valid dot that has just been created. It says so in the comments.

Vagabond commented:
It looks like the code paths where dvv is NOT enabled are not covered by the eunit suite in riak_object anymore. Should we reinstate some tests for the old codepath as well?

Vagabond commented:
It also looks like the ec_eqc test no longer excersises the DVV=false code path, either.

russelldb committed:
Address review comments

Add further code comments
Clean up reconcile to a single fold
Document merge test
Cause ec_eqc to test both dvv and non-dvv paths

russelldb commented:
I've addressed the review comments in the last commit.

Looking at coverage there is still no test for merging CRDTs in riak_object (although the CRDT merge logic is tested elsewhere), even though we are short on time I think we really need this, and maybe even an explicit eqc test for `reconcile/2`, `merge/2`, and `syntactic_merge/2`. I'll let the reviewer decide if that test is needed _before_ we merge this and freeze on Friday, or after.

russelldb committed:
Fix comment typo

Fix ec_eqc to use correct vclock API

seancribbs commented on src/riak_object.erl:
```diff
@@ -182,34 +202,33 @@ ancestors(Objects) ->
+%% to get such a list first.
+reconcile(Objects) ->
+    lists:foldl(fun(O, Acc) ->
+                        syntactic_merge(O, Acc)
```
If you're not changing argument order, this is shorter and clearer:

``` erlang
fun syntactic_merge/2
```

seancribbs commented on priv/riak_kv.schema:
```diff
@@ -337,3 +337,26 @@
+
+{mapping, "dvv", "riak_kv.dvv_enabled", [
+  {default, on},
+  {datatype, {enum, [on, off]}}
```
Please add `{level, advanced}`. For new clusters, we want this behavior on and not directly exposed to new users.

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1292,6 +1290,9 @@ select_newest_content(Mult) ->
 put_merge(false, false, CurObj, UpdObj, _VId, _StartTime) -> % coord=false, LWW=false
+    %% a downstream merge, or replication of a coordinated PUT
+    %% Merge the value received with local replica value
+    %% and store the value IIF it is different to what we already have
```
"IIF" -> "IFF"

seancribbs commented on src/riak_kv_vnode.erl:
```diff
@@ -1301,21 +1302,28 @@ put_merge(false, false, CurObj, UpdObj, _VId, _StartTime) -> % coord=false, LWW=
+            MergedClock = vclock:merge([PutVC, LocalVC]),
+            FrontierClock = vclock:increment(VId, StartTime, MergedClock),
+            {ok, Dot} = vclock:get_entry(VId, FrontierClock),
+            %% Asign an event to the put value
```
"Assign"

seancribbs commented on src/riak_object.erl:
```diff
@@ -239,26 +258,187 @@ compare_content_dates(C1,C2) ->
+                Object#r_object.contents).
+
+%% @private called once for each content entry for each object being
+%% merged. It's job is to drop siblings that are causally dominated,
```
It's / Its

seancribbs commented on src/riak_object.erl:
```diff
@@ -239,26 +258,187 @@ compare_content_dates(C1,C2) ->
+        {ok, Dot} ->
+            case {vclock:descends(Clock, [Dot]), orddict:is_key(Dot, Drop)} of
+                {true, true} ->
+                    %% When the exact same dot present is present in
```
"present" twice

seancribbs commented on src/riak_object.erl:
```diff
@@ -239,26 +258,187 @@ compare_content_dates(C1,C2) ->
+                    Content = convert_to_dict_list(C0),
+                    MergeAcc#merge_acc{keep=ordsets:add_element(Content, Keep)};
+                {true, false} ->
+                    %% The `dot' os dominated by the other object's
```
"os" / "is"

seancribbs commented on test/ec_eqc.erl:
```diff
@@ -791,7 +802,7 @@ kv_vnode(#msg{from = From, c = {coord_put, ReqId, NewObj, Timestamp}},
     WMsg = new_msg(Name, From, {w, Idx, ReqId}, Pri1),
     NodeId = {vc, Start, Node},
-    NewObj1 = riak_object:increment_vclock(NewObj, NodeId, ReqId),
+    NewObj1 = NewObj,%%riak_object:increment_vclock(NewObj, NodeId, ReqId),
```
Rebinding this is kind of pointless, why not just rename the variable through the rest of the function?

seancribbs commented on test/ec_eqc.erl:
```diff
@@ -882,21 +893,30 @@ syntactic_put_merge(CurObj, UpdObj) ->
-    UpdVC = riak_object:vclock(UpdObj),
+coord_put_merge(undefined, UpdObj, NodeId, Timestamp) ->
+    riak_object:increment_vclock(UpdObj, NodeId, Timestamp);
+coord_put_merge(LocalObj, PutObj, NodeId, Timestamp) ->
```
I guess this is ok, but it feels like it mirrors the actual code too much.

russelldb replied:
Man, believe me when I say "I know." But do you really think I should re-write `ec_eqc` for this change? Jon and I discussed this, there are "issues" with this test in general. But this is not the time. If you feel strongly, open an issue, please.

seancribbs replied:
Yeah, it wasn't so much a "this has to change for this PR to land" but more like "ugh".

seancribbs commented:
Some dialyzer warnings related to these changes that are worth fixing:

```
riak_object.erl:336: The call vclock:descends(Clock::vclock:vclock(),[vclock:dot(),...]) contains an opaque term as 2nd argument when terms of different types are expected in these positions
```

This seems to be the construction of a `vclock()` type from a `dot()` outside of the `vclock` module. (Note both `vclock()` and `dot()` are opaque.) vclock should have an exported function which is specifically for dots, or constructs a minimal clock from a dot.

```
riak_object.erl:445: Function convert_to_dict_list/1 has no local return
riak_object.erl:446: Record construction #r_content{metadata::[{_,_}]} violates the declared type of field metadata::'undefined' | dict()
riak_object.erl:449: Function convert_from_dict_list/1 has no local return
riak_object.erl:450: The call dict:from_list(Dict::'undefined' | dict()) contains an opaque term as 1st argument when a structured term of type [{_,_}] is expected
```

Ignore the "no local return" errors, this is trying to construct a dict from the atom `undefined` in some cases. Not necessarily your fault, but it is called by your changes.

```
riak_object.erl:597: Type specification riak_object:dot_key() -> binary() is a supertype of the success typing: riak_object:dot_key() -> <<_:24>>
```

Can this function just be a macro, avoiding a function call where it is used?

Aside from those dialyzer issues and the other things Andrew and I commented on, I'm :+1: to this change. Tests pass, code size is decreased and simplified overall.

Vagabond commented:
I'm generally +1 on this, I asked Sean to be a second reviewer, time permitting, so once you make Sean happy with this too, I think you're cleared for merge.

russelldb commented:
Will fix the dialyzer problems, but please be aware that _all_ but the `dot` issues were added by some other code typist, m'kay? (I agree that it is right to fix these, but better still would be consistent review so one code typist didn't need to clean up after another as part of their general work.) I find your reviews to be amongst the most thorough (talking to you @seancribbs), can you jot down a list of how you perform a review, please? I think it would be valuable.

russelldb commented:
@seancribbs with regard to 

```
riak_object.erl:597: Type specification riak_object:dot_key() -> binary() is a supertype of the success typing:    riak_object:dot_key() -> <<_:24>>
```

> > Can this function just be a macro, avoiding a function call where it is used?

Do you mean a macro that is shared outside riak_object?

russelldb committed:
Address review comments

Fix up dialyzer, move dot macro to common hrl etc

russelldb commented:
Hopefully that covers it. Thanks for the thorough review, gentlemen.

seancribbs commented:
@rdb I didn't realize the Dot key had to be shared outside riak_object. Where else is it called?

seancribbs commented:
@russelldb I'll send you some notes. I'm not sure I've codified it, really, but there are some typical things I do.

russelldb commented:
@seancribbs the ?DOT is used in riak_kv_crdt. CRDTs drop the dot.

seancribbs commented:
@russelldb I see now, thanks!

russelldb committed:
Stylistic cleanup as suggested by Sean Cribbs

Some white space cleanup from last commit

russelldb commented:
@Vagabond @seancribbs happy?

seancribbs commented:
@russelldb :+1:

Vagabond commented:
+1
